python/ray/data/tests/test_dataset_formats.py

```diff
@@ -750,7 +748,7 @@ def test_numpy_read(ray_start_regular_shared, tmp_path):
     np.save(os.path.join(path, "test.npy"), np.expand_dims(np.arange(0, 10), 1))
     ds = ray.data.read_numpy(path)
     assert str(ds) == (
-        "Dataset(num_blocks=1, num_rows=None, "
+        "Dataset(num_blocks=1, num_rows=10, "
```

Comment:
Hmm, how are we able to get the number of rows here without eager reading?

Reply:
AFAICT, before this PR, we were already eagerly submitting the first read task AND blocking on the read task finishing in order to fetch the schema in this case ([1](https://github.com/ray-project/ray/blob/a051e693c106adeb3ae3351bce505be98ca6b89b/python/ray/data/read_api.py#L299-L301), [2](https://github.com/ray-project/ray/blob/a051e693c106adeb3ae3351bce505be98ca6b89b/python/ray/data/impl/block_list.py#L186-L204)). After this PR, we [fetch all of the block metadata](https://github.com/ray-project/ray/blob/d9b28911f87913b4b7e1f41cae148f6400c29afd/python/ray/data/impl/lazy_block_list.py#L312-L325) so we can also get an more informative view of e.g. the number of rows for datasources lacking metadata peeking, as in this case.

Looking at this code again, it gave me an idea for unifying a bit of the metadata code between `BlockList` and `LazyBlockList`, let me make a few small changes that should simplify this a bit.

Reply:
The code linked seems still just look at the metadata for the first block? Also the above "test_numpy_roundtrip" test still has num_rows=None.

Reply:
>The code linked seems still just look at the metadata for the first block?

@jianoaix That's an older commit in this PR, as you can see in the [current state](https://github.com/ray-project/ray/blob/dade52205d1cfaf182f775979fdeedfddc55b448/python/ray/data/impl/lazy_block_list.py#L319), we block and fetch the metadata for the first block if the schema isn't set.

>Also the above "test_numpy_roundtrip" test still has num_rows=None.

The reasoning for that is because the dataset for the `test_numpy_read` dataset only consists of a single block, so our blocking and fetching the metadata for the first block gives us the exact number of rows in the Dataset, and we can calculate the number of rows off of the block metadata alone. The `test_numpy_roundtrip` dataset, on the other hand, has 2 blocks, so when we block and fetch the metadata for the first block, we end up having the number of rows for only one of the two blocks, so we can't calculate the number of rows in the full dataset via the block metadata alone.